### ISSUE TITLE
Postgresql 16 compatibility (with new roles management)

### ISSUE THREAD

**Issue by hargut** [cyrilgdn/terraform-provider-postgresql#407]
### Affected Resource(s)
- postgresql_database
- postgresql_role

provider version: `1.21.0`

The issue is likely related to Postgres 16 Permission/Grant changes.

### Terraform Configuration Files
```hcl
resource "postgresql_role" "dbs" {
  for_each = var.dbs
  name     = each.key
  login    = true
  password = random_password.dbs[each.key].result
}

resource "postgresql_database" "dbs" {
  for_each = var.dbs
  name              = each.key
  owner             = postgresql_role.dbs[each.key].name
  lc_collate        = "en_US.utf8"
  allow_connections = true
}
```

### Expected Behavior
Database should be created.

### Actual Behavior
`Error: Error creating database "<name>": pq: must be able to SET ROLE "<name>"`

### Postgres 16 Release and Documentation References
Details related to permission in Postgres 16 Changelog:
https://www.postgresql.org/docs/16/release-16.html
> Add [GRANT](https://www.postgresql.org/docs/16/sql-grant.html) to control permission to use [SET ROLE](https://www.postgresql.org/docs/16/sql-set-role.html) (Robert Haas)
>
> This is controlled by a new GRANT ... SET option.

https://www.postgresql.fastware.com/blog/discussing-postgresql-what-changes-in-version-16
> The GRANT … SET option was added
>
> The SET option, if set to TRUE, allows the member to change to the granted role using the SET ROLE command. To create an object owned by another role or to give ownership of an existing object to another role, you must have the ability to SET ROLE to that role. Otherwise, commands such as ALTER ... OWNER TO or CREATE DATABASE ... OWNER will fail.

https://www.postgresql.org/docs/current/sql-grant.html
Section:
> GRANT on Roles

It looks like that this permission is not applied to the db admin user when creating the role.
Assume changes would be needed in the `postgresql_role` to be able to grant this option to the according user or implicitly always grant it to the creating user (provider connection).

### Important Factoids
Azure PostgreSQL flexible server version 16

**Comment by cyrilgdn** [comment]
Hey @hargut ,

Thanks for opening the issue. I'm able to  reproduce this problem. 
Indeed, apparently there's many changes in PG 16 about roles.

It's mainly linked to:  https://www.postgresql.org/docs/current/runtime-config-client.html#GUC-CREATEROLE-SELF-GRANT

Even with this option disable (default value), the role will now appear in anyway `pg_auth_members` that the provider is checking but with all flag, but the `admin` one (because as mentioned by the doc above `Since a CREATEROLE user is always implicitly granted ADMIN OPTION on created roles`,  disabled)

So as a quick fix we need to adapt the requests that check the members, but these new features will probably imply more changes :scream: (it could simplify a lot some part of the provider though, as managing the case of non-superuser admin like in AWS/Azure/GCP is the biggest pain point in this provider :sweat_smile: )

I allow myself to rename the issue to Postrges v16 compatibility so I can pin it on top.

**Comment by chelseadole** [comment]
Hey, is there any movement on this issue?

**Comment by leonrob** [comment]
also interested in this issue

**Comment by manneymc** [comment]
just ran into this this morning... when i run my terraform with a **<v16**  _Azure Database for PostgreSQL flexible server_, i can create a postgresql_role and assign it as the owner of a postgresql_database

if i run it with a v16 database service i get the following terraform error:

`postgresql_database.keycloak_db: Creating...

 Error: Error creating database "test": pq: must be able to SET ROLE "testDbo"
`

**Comment by cuongquach** [comment]
Hi @cyrilgdn ,

Would be nice if we know some plan about provider `postgresql` will support Postgresql 16 compatible version ?

Thanks so much.

**Comment by LukaSvastVolue** [comment]
@cyrilgdn what is the status of this issue, I see that you unpinned it so it must mean something has changed?

**Comment by leonrob** [comment]
> @cyrilgdn what is the status of this issue, I see that you unpinned it so it must mean something has changed?

@cyrilgdn I would also like to know this.

**Comment by kunalmehta-eve** [comment]
@cyrilgdn  Please do let us know if is there any update on this ?

**Comment by librucha** [comment]
> We also run into this issue. We use a non-superuser admin (on Azure). As a workaround, until the provider is updated, you can manually set default grant options for new roles:
> 
> `ALTER ROLE "admin_role" SET createrole_self_grant = 'set, inherit';`
> 
> I could not find a way to do this with the terraform provider, but you only need to do it once per server / admin user. You can also set this a server parameter, if you have access to that. Azure does not let you do this yes, so the next best thing for us was to set it on the admin user.
> 
> admin_role will then automatically be granted permissions to alter roles it has created, and inherit permissions of that role. If you have roles that where aready created, you might have to re-grant them:
> 
> `GRANT "user_role" TO "admin_role";`
> 
> But new roles will work fine without this.
> 
> See also: http://rhaas.blogspot.com/2023/01/surviving-without-superuser-coming-to.html

I tried it on Azure Database for PostgreSQL - Flexible Server and is not working.
New roles are still created without non admin membership.

But most strange is if you try create role with REPLIACTION option. New role will be created without any mebership to creator. My discussion about it is here https://learn.microsoft.com/en-us/answers/questions/2084617/azure-postgresql-16-create-role-replication-lacks

**Comment by librucha** [comment]
I'm working on fix this issue ...

**Comment by librucha** [comment]
Hi @cyrilgdn 
I have implemented solution for "fix" this behavior without change parameters `createrole_self_grant` on role or database but here is the questuion is it a good aproach?
Developers of postgresql made this restriction for more granular security settings.

Maybe it can be configurable in providet settings.
What do you mean about it?

As second solution I added `createrole_self_grant` attribute to `postgresql_role` for controling roles created by terraform and fixed test for ne role behavior in PG 16

**Comment by librucha** [comment]
My changes are published here https://registry.terraform.io/providers/librucha/postgresql/latest

**Comment by cyrilgdn** [comment]
Hi,

I've just released a v1.24 with a small fix to make the provider work without changing the `createrole_self_grant` setting.
This fixes the `pq: must be able to SET ROLE` error. Let me know if it's not enough.

For the `createrole_self_grant` setting we'll create a new release a bit later with the work done by @librucha in #481 
I close this issue but I'll create a new one for that.
